server.js

```diff
@@ -1,23 +1,180 @@
-const messages = [welcomeMessage];
+// const messages = [welcomeMessage];
+const messages =[
+  {
+      id: 0,
+      from: "Bart",
+      text: "Welcome to CYF chat system!"
+    },
+    {
+      id: 1,
+      from: "Bart",
+      text: "I hope you enjoy this course"
+    },
+    {id:2,
+    from:"farhana",
+    text:"hello"},
+    {id:3,
+    from:"humail",
+    text:"bye"},
+    {
+      id: 4,
+      from: "Bart",
+      text: "Welcome to CYF chat system!"
+    },
+    {
+      id: 5,
+      from: "Bart",
+      text: "I hope you enjoy this course"
+    },
+    {id:6,
+    from:"farhana",
+    text:"hello"},
+    {id:7,
+    from:"humail",
+    text:"bye"},
+    {
+      id: 8,
+      from: "Bart",
+      text: "Welcome to CYF chat system!"
+    },
+    {
+      id: 9,
+      from: "Bart",
+      text: "I hope you enjoy this course"
+    },
+    {id:10,
+    from:"farhana",
+    text:"hello"},
+    {id:11,
+    from:"humail",
+    text:"bye"},
+    {
+      id:12,
+      from: "dania",
+      text: "call back"
+    },
+    {
+      id:13,
+      from:"fatima",
+      text:"sorry busy rn!"
+    },
+
+]
+    
 
-app.get("/", function (request, response) {
+
+app.get("/", function(request, response) {
   response.sendFile(__dirname + "/index.html");
 });
 
-app.listen(process.env.PORT);
+//Post message
+app.post("/messages", function(request, response) {
+  if (request.body.from === "" || request.body.text === "") {
+    response.status(400).json({
+      message: "Information Missing"
+    });
+  } else {
+    let newId = messages[messages.length - 1].id + 1;
+    const newMessage = {
+      id: newId,
+      from: request.body.from,
+      text: request.body.text
+    };
+    messages.push(newMessage);
+
+    response.status(201).json({
+      message: "message sent"
+      //messageContent: newMessage
+    });
+  }
+});
+
+//Get all messages and check for order and limit
+app.get("/messages", function(request, response) {
+  
+  const order = request.query.order;
+  const limit = request.query.limit;
+  if(order === 'desc' && limit == 10){
+    const descOrder = messages.reverse();
+    const orderedAndLimited = descOrder.slice(0,10);
+    response.json(orderedAndLimited);
+  }else{
+    response.json(messages);    
+  }
+  
+});
+
+//get message by Id
+app.get("/messages/:messageId", (request, response) => {
+  const { messageId } = request.params;
+  const message = messages.find(m => m.id === parseInt(messageId));
+  response.json(message);
+});
+
+//delete message by Id
+app.delete("/messages/:messageId", (request, response) => {
+  const { messageId } = request.params;
+  let index = messages.findIndex(message => message.id === messageId);
+  const messages = messages.filter(
+    message => message.id === parseInt(messageId)
+  );
+  if (messages.length !== 0) {
+    messages.splice(index, 1);
+    response.status(200).send();
+  } else {
+    response.status(400).send();
+  }
+});
```

Comment:
This endpoint is't working at the moment when I run it. I think there's a couple of things which can be fixed:

### Fix 1
The task here is to take a `messageId`, apply that against the array of messages, and where that `messageID` matches the `id` value of an item within that array, we remove it from the array.

There's a few ways that you can go about this, but I think the easiest, is one that you've already started to apply, which is the array method `filter`.

Within your `filter` you are saying, "look at each item in the array I'm giving you and where the condition `message.id === parseInt(messageId)` evaluates to true, don't pass that item into the array that the `filter` is creating.

This mechanism alone is sufficient to identify whether or not we're passed a valid `messageID` from the delete request and it's sufficient enough to return an array without the message to be deleted if that `messageId` does exists (the length of the returning array will be less than the length of the input messages array.

That being the case, we don't need the `findIndex` method here because the filter is able to find the message item by comparing id values as it iterates through the messages array.

### Fix 2
Currently your filter is failing with a `referenceError`. The filter method returns a brand new array and not a mutation of the input array. Therefore you can't call it 'messages'. Currently, as your filter is iterating through messages one item at  a time, it's trying to rewrite onto the messages array.